File: 04-structural-patterns/02-adapter/02-adapter-factory.py
# === Common Interface ===
class PaymentProcessor:
    def pay(self, amount):
        raise NotImplementedError()

# === Adaptees ===
class PayPalService:
    def send_payment(self, amount):
        print(f"PayPal: {amount} paid")

class StripeService:
    def charge_card(self, amount):
        print(f"Stripe: {amount} charged")

class RazorpayService:
    def make_transaction(self, amount):
        print(f"Razorpay: {amount} transacted")

# === Adapters ===
class PayPalAdapter(PaymentProcessor):
    def __init__(self, service):
        self.service = service

    def pay(self, amount):
        self.service.send_payment(amount)

class StripeAdapter(PaymentProcessor):
    def __init__(self, service):
        self.service = service

    def pay(self, amount):
        self.service.charge_card(amount)

class RazorpayAdapter(PaymentProcessor):
    def __init__(self, service):
        self.service = service

    def pay(self, amount):
        self.service.make_transaction(amount)

# === Factory ===
class PaymentGatewayFactory:
    @staticmethod
    def get_processor(gateway: str) -> PaymentProcessor:
        match gateway.lower():
            case "paypal":
                return PayPalAdapter(PayPalService())
            case "stripe":
                return StripeAdapter(StripeService())
            case "razorpay":
                return RazorpayAdapter(RazorpayService())
            case _:
                raise ValueError("Unsupported payment gateway")

# === Client ===
def checkout(amount, gateway_name):
    processor = PaymentGatewayFactory.get_processor(gateway_name)
    processor.pay(amount)

# === Usage ===
checkout(1000, "paypal")
checkout(2000, "stripe")
checkout(3000, "razorpay")

File: 04-structural-patterns/02-adapter/02-adapter-factory.go
package main

import "fmt"

// Common interface
type PaymentProcessor interface {
	Pay(amount float64)
}

// Adaptees
type PayPalService struct{}

func (p *PayPalService) SendPayment(amount float64) {
	fmt.Printf("PayPal: %.2f paid\n", amount)
}

type StripeService struct{}

func (s *StripeService) ChargeCard(amount float64) {
	fmt.Printf("Stripe: %.2f charged\n", amount)
}

type RazorpayService struct{}

func (r *RazorpayService) MakeTransaction(amount float64) {
	fmt.Printf("Razorpay: %.2f transacted\n", amount)
}

// Adapters
type PayPalAdapter struct{ svc *PayPalService }

func (a *PayPalAdapter) Pay(amount float64) {
	a.svc.SendPayment(amount)
}

type StripeAdapter struct{ svc *StripeService }

func (a *StripeAdapter) Pay(amount float64) {
	a.svc.ChargeCard(amount)
}

type RazorpayAdapter struct{ svc *RazorpayService }

func (a *RazorpayAdapter) Pay(amount float64) {
	a.svc.MakeTransaction(amount)
}

// Factory
func GetPaymentProcessor(gateway string) (PaymentProcessor, error) {
	switch gateway {
	case "paypal":
		return &PayPalAdapter{&PayPalService{}}, nil
	case "stripe":
		return &StripeAdapter{&StripeService{}}, nil
	case "razorpay":
		return &RazorpayAdapter{&RazorpayService{}}, nil
	default:
		return nil, fmt.Errorf("unsupported payment gateway: %s", gateway)
	}
}

// Client
func Checkout(amount float64, gateway string) {
	processor, err := GetPaymentProcessor(gateway)
	if err != nil {
		fmt.Println(err)
		return
	}
	processor.Pay(amount)
}

func main() {
	Checkout(1000, "paypal")
	Checkout(2000, "stripe")
	Checkout(3000, "razorpay")
}
